File: src/lib/producer/contracts/guided-media-profile.ts
/** Distinct private manual-crop class, not raw-request interpretation or visual approval. */
import { objectValue } from "./validation";
import { CAPTION_PROFILE, CAPTION_SHORT_PROFILE, CAPTION_BODY_PROFILE, CAPTION_SHORT_BODY_PROFILE,
  SCREENED_CAPTION_PROFILE, SCREENED_CAPTION_SHORT_PROFILE, SCREENED_CAPTION_BODY_PROFILE,
  SCREENED_CAPTION_SHORT_BODY_PROFILE, assertCaptionPresetPlan } from "./guided-caption-profile";

export const MANUAL_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-v1" as const;
export const MANUAL_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-body-v1" as const;
const BODY_BY_OPENING = { "unity-source-float-own-screen-v1": "held-source-float-own-screen-body-v1",
  [MANUAL_SHORT_PROFILE]: MANUAL_SHORT_BODY_PROFILE, [CAPTION_PROFILE]: CAPTION_BODY_PROFILE,
  [CAPTION_SHORT_PROFILE]: CAPTION_SHORT_BODY_PROFILE, [SCREENED_CAPTION_PROFILE]: SCREENED_CAPTION_BODY_PROFILE,
  [SCREENED_CAPTION_SHORT_PROFILE]: SCREENED_CAPTION_SHORT_BODY_PROFILE } as const;
export type OpeningMediaProfile = keyof typeof BODY_BY_OPENING;
export type BodyMediaProfile = typeof BODY_BY_OPENING[OpeningMediaProfile];

export function openingMediaProfile(value: unknown): OpeningMediaProfile {
  if (typeof value !== "string" || !Object.hasOwn(BODY_BY_OPENING, value)) {
    throw new Error("Private opening media profile is unsupported");
  }
  return value as OpeningMediaProfile;
}

export function bodyMediaProfile(value: OpeningMediaProfile): BodyMediaProfile {
  return BODY_BY_OPENING[openingMediaProfile(value)];
}

export function isManualMediaProfile(value: unknown): boolean {
  return value === MANUAL_SHORT_PROFILE || value === CAPTION_SHORT_PROFILE || value === SCREENED_CAPTION_SHORT_PROFILE;
}

function active(value: unknown): boolean {
  if (Array.isArray(value)) return value.length > 0;
  if (value !== null && typeof value === "object") return Object.keys(value).length > 0;
  return Boolean(value);
}

/** Property presence is unsupported intent even when its supplied value is empty. */
function assertNoPresenterLayouts(plan: Record<string, unknown>): void {
  if (Object.hasOwn(plan, "presenterLayouts")) throw new Error("Legacy opening profile has no presenterLayouts execution owner");
}

/** Exact explicit subset of plan_lint_reframe; actual displayed geometry remains Python-observed. */
export function assertManualShortPlan(plan: Record<string, unknown>): void {
  assertNoPresenterLayouts(plan); assertManualShortGeometry(plan, false);
}

export function assertManualCaptionShortPlan(plan: Record<string, unknown>): void {
  assertCaptionPresetPlan(plan); assertNoPresenterLayouts(plan); assertManualShortGeometry(plan, true);
}

/** Shared declared crop subset only, not profile admission. Public legacy entries retain their owner fence. */
export function assertManualShortGeometry(plan: Record<string, unknown>, captioned: boolean): void {
  const target = objectValue(plan.target, "manual short target"), reframe = objectValue(plan.reframe, "manual short reframe");
  if (target.mode !== "short" || target.width !== 1080 || target.height !== 1920) {
    throw new Error("Manual short profile requires explicit1080x1920 short target");
  }
  if (Object.keys(reframe).sort().join(",") !== "crop,layout,track" || reframe.layout !== "fill" || reframe.track !== false) {
    throw new Error("Manual short requires only explicit fill/crop/track:false");
  }
  if (objectValue(plan.captions, "manual short captions").burn !== captioned) {
    throw new Error("Manual short requires explicit captions.burn=false; captioned short is unqualified");
  }
  if ([plan.overlays, plan.presenter].some(active)) {
    throw new Error("Manual short has unsupported additional visual intent");
  }
  const crop = reframe.crop;
  if (!Array.isArray(crop) || crop.length !== 4 || crop.some((item) => typeof item !== "number" || !Number.isFinite(item))) {
    throw new Error("Manual short crop requires four finite nonboolean numbers");
  }
  const [x, y, width, height] = crop;
  // Existing REFRAME_SPLIT.crop_min_frac, parity-tested against Python; no changed renderer bound.
  if (!(x >= 0 && y >= 0 && x <= 1 && y <= 1 && width >= 0.05 && height >= 0.05 && x + width <= 1 && y + height <= 1)) {
    throw new Error("Manual short crop is outside the existing normalized source bounds");
  }
  if (!Array.isArray(plan.cutTrack) || !plan.cutTrack.length) throw new Error("Manual short needs an explicit nonempty cut track");
  const cuts = plan.cutTrack.map((value) => objectValue(value, "manual short cut")), source = cuts[0].sourceId;
  if (typeof source !== "string" || !source || cuts.some((row) => row.sourceId !== source)) {
    throw new Error("Manual short supports exactly one used source and one fixed crop");
  }
}

function freshProfileForPlan(plan: Record<string, unknown>): OpeningMediaProfile {
  const target = objectValue(plan.target, "opening target"), reframe = plan.reframe;
  if (plan.captionsTrack && typeof plan.captionsTrack === "object" && !Array.isArray(plan.captionsTrack)) {
    assertCaptionPresetPlan(plan);
    if (target.mode === "short") { assertManualCaptionShortPlan(plan); return SCREENED_CAPTION_SHORT_PROFILE; }
    return SCREENED_CAPTION_PROFILE;
  }
  if (target.mode === "short" && reframe && typeof reframe === "object" && Object.hasOwn(reframe, "crop")) {
    assertManualShortPlan(plan);
    return MANUAL_SHORT_PROFILE;
  }
  return "unity-source-float-own-screen-v1";
}

/** Fresh execution requires screening. A separately held legacy input keeps its exact original class.
 * The optional token is not public request authority: callers still verify original input/authority bytes. */
export function openingMediaProfileForPlan(plan: Record<string, unknown>, heldProfile?: unknown): OpeningMediaProfile {
  assertNoPresenterLayouts(plan);
  const selected = freshProfileForPlan(plan);
  if (heldProfile === undefined) return selected;
  const held = openingMediaProfile(heldProfile);
  const legacy = (selected === SCREENED_CAPTION_PROFILE && held === CAPTION_PROFILE)
    || (selected === SCREENED_CAPTION_SHORT_PROFILE && held === CAPTION_SHORT_PROFILE);
  if (held !== selected && !legacy) throw new Error("Held opening profile differs from the exact candidate class");
  return held;
}

File: scripts/producer/guided_media_profile.py
"""Versioned private media classes; explicit geometry is not framing approval.

The historical class is unchanged. The new class executes only a submitted
manual fill crop through ordinary reframe, never interprets a raw request or
enables captions, tracking, grading or other finishing lanes.
"""
from __future__ import annotations

import math

from producer_config import REFRAME_SPLIT
from guided_caption_profile import (CAPTION_PROFILE, CAPTION_SHORT_PROFILE, CAPTION_BODY_PROFILE,
    CAPTION_SHORT_BODY_PROFILE, SCREENED_CAPTION_PROFILE, SCREENED_CAPTION_SHORT_PROFILE,
    SCREENED_CAPTION_BODY_PROFILE, SCREENED_CAPTION_SHORT_BODY_PROFILE, caption_preset_plan)

OPENING_PROFILE = "unity-source-float-own-screen-v1"
SHORT_PROFILE = "unity-source-float-manual-short-own-screen-v1"
BODY_PROFILE = "held-source-float-own-screen-body-v1"
SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-body-v1"
_OMITTED_PROFILE = object()


def opening_profile(value: object) -> str:
    """Parse a known execution class without claiming its plan is admissible."""
    if type(value) is not str or value not in (OPENING_PROFILE, SHORT_PROFILE, CAPTION_PROFILE, CAPTION_SHORT_PROFILE,
                                               SCREENED_CAPTION_PROFILE, SCREENED_CAPTION_SHORT_PROFILE):
        raise RuntimeError("private opening media profile is unsupported")
    return value


def body_profile(value: object) -> str:
    """Preserve a one-to-one opening/body class relation, never upgrade a result."""
    return {OPENING_PROFILE: BODY_PROFILE, SHORT_PROFILE: SHORT_BODY_PROFILE,
            CAPTION_PROFILE: CAPTION_BODY_PROFILE, CAPTION_SHORT_PROFILE: CAPTION_SHORT_BODY_PROFILE,
            SCREENED_CAPTION_PROFILE: SCREENED_CAPTION_BODY_PROFILE,
            SCREENED_CAPTION_SHORT_PROFILE: SCREENED_CAPTION_SHORT_BODY_PROFILE}[opening_profile(value)]


def manual_profile(value: object) -> bool:
    """Both explicit manual classes share the same actually proved crop geometry."""
    return type(value) is str and value in (SHORT_PROFILE, CAPTION_SHORT_PROFILE, SCREENED_CAPTION_SHORT_PROFILE)


def assert_no_presenter_layouts(plan: dict) -> None:
    """Reject present unsupported intent, including null, empty and false values."""
    if "presenterLayouts" in plan:
        raise RuntimeError("legacy opening profile has no presenterLayouts execution owner")


def manual_short_plan(plan: dict) -> dict:
    """Require exact current-candidate geometry and explicit captions-off intent."""
    return _manual_short_plan(plan, False)


def manual_caption_short_plan(plan: dict) -> dict:
    """Require unchanged manual geometry plus explicit qualified caption preset."""
    caption_preset_plan(plan)
    return _manual_short_plan(plan, True)


def _manual_short_plan(plan: dict, captioned: bool) -> dict:
    """Shared crop checks; the ordinary prepared BASE remains caption-free."""
    assert_no_presenter_layouts(plan)
    return manual_short_geometry(plan, captioned)


def manual_short_geometry(plan: dict, captioned: bool) -> dict:
    """Geometry validation only; callers separately enforce their exact lane/profile.

    This shared helper grants no presenter execution capability and removes no
    candidate field. The historical public profile validators retain their fence.
    """
    target, reframe, captions = (plan.get(key) for key in ("target", "reframe", "captions"))
    if type(target) is not dict or target.get("mode") != "short" \
            or type(target.get("width")) is not int or target["width"] != 1080 \
            or type(target.get("height")) is not int or target["height"] != 1920:
        raise RuntimeError("manual short profile requires explicit1080x1920 short target")
    if type(reframe) is not dict or set(reframe) != {"layout", "crop", "track"} \
            or reframe["layout"] != "fill" or reframe["track"] is not False:
        raise RuntimeError("manual short requires only explicit fill/crop/track:false")
    if type(captions) is not dict or captions.get("burn") is not captioned:
        raise RuntimeError("manual short requires explicit captions.burn=false; captioned short is unqualified")
    if any(plan.get(key) for key in ("overlays", "presenter")):
        raise RuntimeError("manual short has unsupported additional visual intent")
    crop = reframe["crop"]
    if type(crop) is not list or len(crop) != 4 or any(
            type(item) not in (int, float) or not math.isfinite(item) for item in crop):
        raise RuntimeError("manual short crop requires four finite nonboolean numbers")
    x, y, width, height = crop
    if not (0 <= x <= 1 and 0 <= y <= 1 and width >= REFRAME_SPLIT["crop_min_frac"]
            and height >= REFRAME_SPLIT["crop_min_frac"] and x + width <= 1 and y + height <= 1):
        raise RuntimeError("manual short crop is outside the existing normalized source bounds")
    cuts = plan.get("cutTrack")
    if type(cuts) is not list or not cuts or any(type(row) is not dict for row in cuts):
        raise RuntimeError("manual short needs an explicit nonempty cut track")
    source = cuts[0].get("sourceId")
    if type(source) is not str or not source or any(row.get("sourceId") != source for row in cuts):
        raise RuntimeError("manual short supports exactly one used source and one fixed crop")
    return {"sourceId": source, "reframe": reframe, "canvas": [1080, 1920], "captionsBurned": False}


def profile_for_plan(plan: dict, held_profile: object = _OMITTED_PROFILE) -> str:
    """Select fresh screened intent or validate an EXACT held compatible old class.

    The held token is already part of input/authority hashes, never a request to
    downgrade new work. No result is rewritten or upgraded by this selector.
    """
    assert_no_presenter_layouts(plan)
    selected = _fresh_profile(plan)
    if held_profile is _OMITTED_PROFILE:
        return selected
    opening_profile(held_profile)
    compatible = {SCREENED_CAPTION_PROFILE: (SCREENED_CAPTION_PROFILE, CAPTION_PROFILE),
                  SCREENED_CAPTION_SHORT_PROFILE: (SCREENED_CAPTION_SHORT_PROFILE, CAPTION_SHORT_PROFILE)}
    if held_profile not in compatible.get(selected, (selected,)):
        raise RuntimeError("held opening profile is incompatible with the exact candidate plan")
    return held_profile


def _fresh_profile(plan: dict) -> str:
    """Only fresh authored caption plans select the new explicit screened class."""
    target, reframe = plan.get("target") or {}, plan.get("reframe") or {}
    if type(plan.get("captionsTrack")) is dict:
        caption_preset_plan(plan)
        if target["mode"] == "short":
            manual_caption_short_plan(plan)
            return SCREENED_CAPTION_SHORT_PROFILE
        return SCREENED_CAPTION_PROFILE
    if type(target) is dict and target.get("mode") == "short" \
            and type(reframe) is dict and "crop" in reframe:
        manual_short_plan(plan)
        return SHORT_PROFILE
    return OPENING_PROFILE
